### PR TITLE
Kwxm/machines eval builtins

### PR DESCRIPTION
**Pull request by kwxm**
Here's a document about some of the issues that have surfaced wrt evaluation of unsaturated builtins, including Roman's recent point about higher-rank builtins (at the end).  I thought I'd better write some of this down before we forgot what's going on; also I'm not entirely sure how to proceed with the CK machine.  

Feel free to comment/edit/add stuff.

### CODE REVIEW

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+unnecessarily wrap them in a `delay`. [But what happens if the final
+argument is a type argument, ie, if a builtin returns a polymorphic
+value?  We currently require all builtin arguments to be values and
+builtins are supposed to be parametrically polymorphic _and_ return
+values, so can a builtin really return something that can compute when
+instantiated?  Perhaps it can: maybe we can pass `Λt:*.error` (which
+erases to `delay error`) to the identity function?]
+
+Anyway, perhaps we could modify `force` so that it'll act on _any_
+value, doing nothing when it's not wrapped in `delay`; we would then
```
Comment (effectfully) [inline notes/fomega/evaluation/builtin-evaluation.md:261]: This suggests that we may want kind of untyped type checking: we need to check that `force` is only applied to

1. `delay`ed things
2. builtins where they expect types

For instance, take the following builtin:

```
id : all a. a -> a
```

We erase it to

```
id : delay (any -> any)
```

where

```
delay :: * -> *
(->) :: * -> * -> *
any :: *

any ~= delay any | (->) any any
```

and we also have

```
force : delay any -> any
```

so that

```
force Erased.id 1
```

is "well-typed".

Given that the denotational semantics of untyped lambda calculus is exactly `any ~ (any -> any)`, it might make sense to extend that with isorecursive `delay` & `force`, which should allow us to check that we apply `force` only in appropriate places.

Or maybe we don't want to check anything statically. Anyway, I think we don't need `force` to act on any value, only on a `delay`ed one or on a builtin that before being erased was expecting a type argument at that particular position. Such a check may as well happen at runtime.

Once I start implementing an evaluator for Untyped Plutus Core, I'll implement this exact runtime check.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+specification, and there's nowhere to store the argument information.
+I've considered a number of ways to fix this.
+
+ * I initially thought we could store the required information
+   in a frame.  This is how the CK machine worked with saturated
+   builtins, but we can't use the same technique with unsaturated
+   builtins.  In the unsaturated case a partially applied builtin
+   is a perfectly legitimate value and could be returned by a function
+   or supplied as an argument to another function.  This just doesn't
+   fit in with the way that stack frames are used.
```
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:163]: Let me think that through properly.  I'd managed to convince myself it didn't work, but now I'm having doubts again.  I think that proves that I should write something down...

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+   phase have to bear any relation to the term type consumed by the
+   compute phase, so this might be the way to go.
+
+   UPDATE: I've now done this and it seems to perform OK: it may be 1%
+   or so slower than the intiial solution described below.  I added a
+   `CkValue` type that was more or less isomorphic to the usual values
+   except for builtins, where the value includes a count of arguments
+   and lists of the type and term arguments it's received so far.
+   Introducing the new type also required quite a lot of typeclass
+   stuff to make it acceptable to the CAM.  See
```
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:181]: Yeah.  In the earlier version of the CK machine it didn't need these because it was using Terms, which already have suitable instances defined somewhere.  Of course, when I tried to use CkValues instead I started to get terrifying type errors. Looking at it again, there's less stuff required to fix that than I thought.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+Also, our primary evaluator will eventually be untyped, and it'll
+presumably be a CEK machine.  We'll surely still want a typed
+evaluator as well though, so maybe that should be the CEK machine as
+well.  It's maybe tempting to use the typed CEK machine and make it a
+bit more robust by using the `TypeArg | TermArg` thing mentioned above
+to count the arities: that'll be slightly less efficient, but we'll
+only need it for the typed evaluator, where efficiency isn't paramount.
+
+
+## Higher-rank builtins
```
Comment (effectfully) [inline notes/fomega/evaluation/builtin-evaluation.md:230]: I initially was confused when you mentioned "higher-rank" builtins, because I always thought those were about types having a `forall` inside the type of an argument. Here's what I see on [Haskell wiki](https://wiki.haskell.org/Rank-N_types):

> Normal Haskell '98 types are considered Rank-1 types. A Haskell '98 type signature such as
>
> `a -> b -> a`
>
> implies that the type variables are universally quantified like so:
>
> `forall a b. a -> b -> a`
>
> `forall` can be floated out of the right-hand side of `->` if it appears there, so:
>
> `forall a. a -> (forall b. b -> a)`
>
> is also a Rank-1 type because it is equivalent to the previous signature. 
>
> However, a `forall` appearing within the left-hand side of `(->)` cannot be moved up, and therefore forms another level or rank. The type is labeled "Rank-N" where N is the number of foralls which are nested and cannot be merged with a previous one. For example:
>
> `(forall a. a -> a) -> (forall b. b -> b)`
> is a Rank-2 type because the latter `forall` can be moved to the start but the former one cannot. Therefore, there are two levels of universal quantification. 

So is it correct to talk about higher-rank builtins here?

Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:230]: Yeah, now I'm just confused. What do we have then? Can we always move the quantifiers to the front in our case?

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+erases to `delay error`) to the identity function?]
+
+Anyway, perhaps we could modify `force` so that it'll act on _any_
+value, doing nothing when it's not wrapped in `delay`; we would then
+be able to completely erase type arguments to builtins except perhaps
+ones at the very end.  (I'm now wondering what would happen if you
+force a _term_: could that ever happen?)
+
+Roman has a number of other proposals, including erasing type arguments
+of builtins to `()` and prohibiting final quantifiers altogether.
```
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:267]: Now I'm not sure exactly what we mean by a final quantifier.  Are we talking about the case when the final argument is a type argument, so you supply all the term arguments and then you still have to instantiate the builtin before in computes?  I'm finding it hard to see why that would occur.  It's a little confusing because the return type could be a forall, and then you'd have to instantiate the _result_ of the builtin; I don't that's what we're talking about here though.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/fomega/evaluation/builtin-evaluation.md:163]: I think the difference is that a lambda does effectively carry its arity with it: it's always 1. So the frame for application knows what to do just from the function term if it's a lambda.

In our case we *could* reconstruct that: the application frame could deconstruct its function term (e.g. using `termAsPrimIterApp`, look up the corresponding arities, and decide whether it's done or not. But that would be repeating a lot of work again.

[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/fomega/evaluation/builtin-evaluation.md:163]: Looks like what I said here is exactly what you said you were doing below!

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+   list of arguments that it's received so far.  That's probably a
+   really bad idea though, since it'd only be there for the benefit of the
+   CK machine and it's taking us back towards the problems we had with
+   saturated builtins.  Let's not do that.
+ 
+
+I'm not sure what we want to do here.  How much do we care about the
+CK machine these days?  I think we originally had it because it was
+pretty simple and easily formalised, and fitted in nicely with the
+reduction semantics.  Maybe that's not so important now?
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/fomega/evaluation/builtin-evaluation.md:211]: I think we still do, but it's okay if it's inefficient.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 4 above)
Comment (michaelpj) [reply notes/fomega/evaluation/builtin-evaluation.md:230]: Yeah, that does sound right. We're talking about cases where we've got type arguments in the middle, where they could be floated to the left. However for us these are not quite the same, because type arguments delay evaluation!

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/fomega/evaluation/builtin-evaluation.md:267]: Well, consider an error builtin, with type `forall a. a`. This takes precisely one argument, a type argument, and only then evaluates to error. Seems at least somewhat reasonable?

[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:267]: No, I think not. The CAM doesn't know anything about types: you give it a builtin and a bunch of terms and it computes a result and returns it.  It doesn't know anything about types, so it computes at the point when it gets its final term argument.  Now you could return a polymorphic value that's just like any other polymorphic value: you can instantiate it, and at that point it might carry out some more computation _outside the constant application machinery_.  There is a quantifier at the end, but it's in the _return type_ of the builtin.  This sort of thing always confuses me because of the fact that builtins have type schemes that don't correspond one-to-one with types: when you convert a type scheme to a type you just get lots of arrows and if you don't know the arity you can't tell whether the arrows at the end belong to the return type or the input type.  

I think something similar is going on with builtins.  If you have a builtin b with signature `[t₁,...,tₙ]r` and we're talking about a final quantifier, does it belong in `tₙ` or in `r`?  If it's in `r` then b's just returning a polymorphic value (which is perfectly legitimate: b might be the identity function and you've given it an abstraction) and there's no problem about instantiating it.  If the quantifier's in `tₙ`, then what have we got?  I think `tₙ` has to be `∀t:K...`, in which case the final argument's a polymorphic value and it doesn't make sense to talk about instantiating the application. 

Does this make sense, or am I misunderstanding?  If we consider a signature like `[t₁,...,tₙ]r`, how are we interpreting the `t`s?  I'm not sure how to make sense of saying that some t represents a type argument (as opposed to a term). If we're saying that a builtin can take an actual type as an argument, then I don't see how that could work: the CAM ignores types, so they can't have any effect on what goes on inside the builtin, and they can't somehow delay evaluation because nothing happens anyway until you get the final _term_ argument. Even if the final argument was a type, wouldn't the builtin still have to compute when it gets the final term?  If there's any delay, it would have to be handled by the machine, not the CAM.

[Off to lie down: I've got no idea what's going on any more.]

[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/fomega/evaluation/builtin-evaluation.md:267]: > No, I think not. The CAM doesn't know anything about types: you give it a builtin and a bunch of terms and it computes a result and returns it. It doesn't know anything about types, so it computes at the point when it gets its final term argument.

This seems wrong if we're serious about turning type arguments into delays that require a force to get past them. If the last argument isn't a type, then you can get away with just ignoring any forces that are applied (assuming they go in the right place), because in the end you'll have to get all your term arguments. Having a type argument as the last argument just puts this into focus.

So I think the CAM *should* consider type arguments.

(I agree that there's a difference between the quantifier being at the end and it being in the return type. But I think those two things *just are different*. And I don't see why the version where the quantifier is at the end of the type argument list shouldn't work.)

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 4 above)
Comment (effectfully) [reply notes/fomega/evaluation/builtin-evaluation.md:230]: > However for us these are not quite the same, because type arguments delay evaluation!

Whatever, I think "higher-rank types" is an established terminology that doesn't quite fit in our case, but not totally sure.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply notes/fomega/evaluation/builtin-evaluation.md:267]: @kwxm 

> Are we talking about the case when the final argument is a type argument, so you supply all the term arguments and then you still have to instantiate the builtin before in computes?

Ah, sorry, yes, I meant that + silently removing all types from builtin applications (which is what you've convinced me was a bad idea). Prohibiting final quantifiers without silently removing types (but using explicit `force` for example) could work indeed.

@michaelpj 

> Well, consider an error builtin, with type forall a. a. This takes precisely one argument, a type argument, and only then evaluates to error. Seems at least somewhat reasonable?

If we ever have polymorphic built-in types, then also `nil : all a. list a`.

[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:267]: > silently removing all types from builtin applications (which is what you've convinced me was a bad idea).

I'm becoming convinced that it's a good idea!

[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:267]: > So I think the CAM should consider type arguments.

That disturbs me.  I don't like it when type-level constructs influence computation.  I tend to feel that you should just be able to throw all the types away without changing anything, but maybe that's just me being narrow-minded.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:163]: I think trying to use frames for partially-applied builtins in the CK machine really forces you to introduce a new type of value to convert the frame into when you run out of arguments and you're still not fully-applied, and this just takes us back to the version with special values. I've updated the document to reflect that.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+   main phases: "compute" and "return", the former acting on terms and
+   the latter acting on values.  In the CK machine, we've been using
+   values of the type described in the Plutus Core specification,
+   which are terms with a particular syntactic form.  However, I think
+   there's no reason why type of the arguments consumed by the return
+   phase have to bear any relation to the term type consumed by the
+   compute phase, so this might be the way to go.
+
+   UPDATE: I've now done this and it seems to perform OK: it may be 1%
+   or so slower than the intiial solution described below.  I added a
```
[the remark this answers is not in the document]
Comment (jmchapman) [reply notes/fomega/evaluation/builtin-evaluation.md]: Indeed, it would be also be nice if the reference implementations performed type substitutions rather than dropping them.

[the remark this answers is not in the document]
Comment (effectfully) [reply notes/fomega/evaluation/builtin-evaluation.md]: Agreed. @michaelpj WDYT?

[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/fomega/evaluation/builtin-evaluation.md]: I agree with both comments.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply notes/fomega/evaluation/builtin-evaluation.md:157]: > I think trying to use frames for partially-applied builtins in the CK machine really forces you to introduce a new type of value to convert the frame into when you run out of arguments and you're still not fully-applied, and this just takes us back to the version with special values. 

Having

```
s, [(λf:int->int . f (con 1)), _], (builtin addInteger 1 [(con 11]) ▻ ???
```

can we turn the builtin frame into a normal term (`addInteger` applied to `11`) and continue evaluation _in the return mode_? Then we'll eventually need to retraverse the same application and turn the term into the frame once again when we have all arguments in place, but this inefficiency is probably not of a much concern, because a partially applied builtin fed to another function is not a very common case and it'll probably be faster than the current solution anyway.

tl;dr: temporary turn builtin frames into stuck applications (known to be values) when we need that and reprocess them later.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+ 2. `F` is an application of a builtin function to a (possibly empty) list of values.
+
+ 3. `F` is something else (a constant for example): in this case an error occurs.
+
+Step 2 is the one that has to be efficient.  In the original CEK machine, a function
+`termAsPrimIterApp` is called.  This takes a term and checks whether it is a sequence of
+type instantiations and applications with the name of a builtin function at the head, for example
+
+```
+  [{([builtin b) V₁ V₂] T₁} V₃]
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:49]: There seems to be an interleaving of values and types here that I wasn't expecting. I thought this was describing the existing machinery that doesn't allow such interleaving.

(and some mismatched brackets)

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+simple benchmarks involving a lot of builtin applications, this was
+substantially faster than the original CEK machine.  However,
+saturated builtins complicated matters considerably and we decided not
+to pursue them: see [this document](../../../language-plutus-core/docs/Saturatedness.md)
+for a consideration of the issues involved.
+One important factor was that an alternative version of the CEK
+machine allowed us to retain unsaturated builtins but achieve similar
+performance to unsaturated builtins.
+
+### The alternative CEK machine Part of the reason for abandoning
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:83]: I think there should be a line break before "Part..."

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+handling environments: we could have retained the original method
+of builtin evaluation and used a simpler version of the `builtin`
+value below while still improving the handling of environments).
+
+The original machine dealt with closures: pairs `(V,ρ)` with `V` a value and `ρ`
+an environment containing values for the free variables in `V` (and probably
+other variables as well: the new machine appeared attractive because it would
+reduce the number of irrelevant variables stored in environments).
+
+In the new proposal, closures are replaced with a new kind of "value":
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:99]: ... where closures `(M,ρ)` appear under binders.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+```
+V := con cn
+   | abs ɑ (M , ⍴)
+   | lam x (M , ρ)
+   | iwrap A B V
+   | builtin ρ bn count [types] [V]
+```
+(we'll call these things "CEK-values")
+
+These are similar to, but distinct from, the "values" in the Plutus Core
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:110]: I cannot resist nit-picking about the values not reducing further. I think that both version could reduce further if we reduced under binders. I think it's easier to point out that the ordinary "values" have just terms under binders.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 12]
```diff
@@ -0,0 +1,347 @@
+# Efficiently evaluating builtin functions in the CK and CEK machines
+
+(This is a bit impressionistic: it's describing ongoing work and discussions.)
+
+It's important that builtin evaluation should be efficient in our
+Plutus Core evaluators.  There's some machinery known as the constant
+application machinery (CAM for short) which takes the name of a
+builtin and a list of arguments of some unspecified type `term` which
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:8]: that sounds pretty elegant.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmchapman) [reply notes/fomega/evaluation/builtin-evaluation.md:157]: I think the values as described in the plutus core spec are designed for use with saturated builtins. So, we should add a constructor for an unsaturated builtin to the ordinary definition of values. We should update the reduction semantics to use this too. Would it help to put the arity info in there?

I have an Agda implementation of an CK + unsaturated builtins: https://github.com/input-output-hk/plutus/blob/kwxm/machines-eval-builtins/metatheory/Scoped/CK.lagda.md. I am not sure how relevant it is. There is arity information everywhere - in the terms, in values and in the frames. And it is a bit obfuscated as the arity info is a proof that the arity is reasonable.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+   machine and the CEK machine both have two main phases: "compute"
+   and "return", the former acting on terms and the latter acting on
+   values.  In the CK machine, we've been using values of the type
+   described in the Plutus Core specification, which are terms with a
+   particular syntactic form.  However, there's no particular reason why
+   the type of the arguments consumed by the return phase has to bear any
+   relation to the term type consumed by the compute phase, so this
+   might be the way to go.
+
+   UPDATE: I've now done this and it seems to perform OK: it may be 1%
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:213]: I think I am arguing for replacing "Value" with "CKValue".

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+   compare it with the number of arguments; if they match, we hand the
+   application off to the CAM.  This is a bit more efficient than the
+   earlier method because we only call the CAM once, but it still
+   requires repeated calls to the inefficient `termAsPrimIterApp`,
+   which we otherwise don't need.  This is a fairly simple
+   modification of the existing machine, but doesn't look very
+   much like the new CEK machine: the version in proposal 2 is
+   more like the new CEK machine.
+
+ 4. We could modify the `builtin` constructor in the AST to contain a
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:239]: I think this is the version I have formalised in Agda.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmchapman) [reply notes/fomega/evaluation/builtin-evaluation.md:249]: I think it's a good reference to keep around and as we don't care about efficiency I personally would like it to do type substitutions properly so it returns terms with correct type annotations.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+CK machine was its simplicity, but maybe it's not possible to be very
+simple if we deal with builtins properly.
+
+Also, our primary evaluator will eventually be untyped, and it'll
+presumably be a CEK machine.  We'll surely still want a typed
+evaluator as well though, so maybe that should be the CEK machine as
+well.  It's maybe tempting to use the typed CEK machine and make it a
+bit more robust by using the `TypeArg | TermArg` thing mentioned above
+to count the arities: that'll be slightly less efficient, but we'll
+only need it for the typed evaluator, where efficiency isn't paramount.
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:265]: it sounds like we might need something similar in the untyped case too.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+looks like
+```
+  [α₁::K₁,...,αₘ::Kₘ](t₁,...,tₙ)r
+```
+where the `α`s are type variables, the `K`s are kinds, the `t`s are
+the types of the arguments and `r` is the return type.  This maps
+onto the type
+
+```
+   ∀α₁::K₁. ... .∀αₘ::Kₘ. t₁ → ... → tₙ → r
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:282]: there are some funny boxes here, perhaps some characters I don't have?

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 1 above)
Comment (jmchapman) [reply notes/fomega/evaluation/builtin-evaluation.md:322]: ^ That's much sophisticated than I was expecting. Take if then else:
```
ifThenElse : forall a. Bool -> a -> a -> a
```
and an example:
```
[[[{(builtin ifThenElse) (con integer)} (con true)] (con 0)] (con 1)]
```
I thought it would erase to:
```
[[[{force (builtin ifThenElse)} (con true)] (con 0)] (con 1)]
```
i.e. type instantiation gets replaced by `force` and we need to track that there are enough `force`s to match the number of type args, delay doesn't get used.

I will have to think some more about it.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmchapman) [reply notes/fomega/evaluation/builtin-evaluation.md:328]: >That disturbs me. I don't like it when type-level constructs influence computation. I tend to feel that you should just be able to throw all the types away without changing anything, but maybe that's just me being narrow-minded.

I think we've already gone down this road with erasing big lambda/type instantiation to delay/force so we should keep things consistent.

File: notes/fomega/evaluation/builtin-evaluation.md [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+more Haskell into Plutus Core?  It appears that there's not a huge
+amount of extra implementation effort involved though, so they
+wouldn't be too difficult to support.
+
+One thing that worries me a little is how signatures for builtins with
+interleaved type and term arguments  affect the specification.  It
+might be quite tricky just to write down a generic instance of such a
+signature, for example.
+
+What about formalisation? Would this complicate things in the Agda
```
Comment (jmchapman) [inline notes/fomega/evaluation/builtin-evaluation.md:343]: This will require some work definitely. It took me a few iterations to have a practical treatment of saturated non-interleaved builtins.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 1 above)
Comment (effectfully) [reply notes/fomega/evaluation/builtin-evaluation.md:322]: > That's much sophisticated than I was expecting. 

It's an alternative idea that is most likely not going to work out. I agree that we should start with what you describe as it makes perfect sense and is easy enough to implement.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 4 above)
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md]: Now I think that we do have higher-rank builtins whatever we do, because they can certainly accept abstractions as arguments (although maybe they can't do a lot with them), and that lets you put a quantifier to the left of as many arrows as you want in the tree representation of the type, which is what higher-rank types are about.

So yes, whatever we've been talking about, I don't think it is higher-rank types: sorry about that.  I think things are a bit confused because again we're talking about multi-argument functions and they're not something that's commonly discussed in type theory.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 8 above)
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:49]: Ah, that's mostly what we're arguing about.  The question is whether we should allow interleaving or not, and the opinion seems to be that we should.  The problem then is what we erase type application to.  We erase `abs` to `delay` and `{}`to `force`, and then we have to ask what happens to instantiations of partially-applied builtins.  A builtin can't compute at all until it gets its final argument, so the only case where there might be a problem is if we have an instantiation after the final term argument.  It's not clear (to me, anyway) if we'd ever want that case though.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 13 above)
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:213]: > I think the conversation is evolving in several different places so apologies for chiming in with comments on already covered stuff.

That's fine: it's definitely quite difficult to follow what's going on though with this non-linear conversation and further comments in the PRs for the machines.

**Comment by kwxm** [comment]
> I think the conversation is evolving in several different places so apologies for chiming in with comments on already covered stuff.

That's fine: it's definitely quite difficult to follow what's going on though with this non-linear conversation and further comments in the PRs for the machines.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md:157]: Hmmm.  I'll think about that.  The version in `kwxm/alternative-ck` just uses frames though, and it has the advantage of being similar to the current version of the alternative CEK machine.  On the ogther hand, if this works maybe we could use it for the CEK machine as well.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 8 above)
Comment (jmchapman) [reply notes/fomega/evaluation/builtin-evaluation.md]: I think returning something of type `forall a . t` from a builtin should be handled by the ordinary erasure machinery as the builtin is no-longer in the picture. But, I could be wrong.

**Comment by kwxm** [comment]
I'm going to merge this so that other people can have a look.  Feel free to keep on commenting.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 8 above)
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md]: That's the thing I'm not sure about either.  Is there any difference between a builtin which consumes all of its term arguments but doesn't compute until it gets instantiated and one which consumes its arguments and returns an abstraction which you can then instantiate?  The instantiation definitely has to be handled by the evaluator because we're not actually passing types to the CAM (and if we did it would probably be unpleasant: we'd have to pass elements of a sum type and the machinery would have to discriminate on those, which would be inefficient.  Maybe we could pass units, but then how would you tell them apart from units passed as term arguments?  Would you need to?)

Comment (jmchapman) [reply notes/fomega/evaluation/builtin-evaluation.md]: Ah, I guess was expecting that the CAM would have to deal with this somehow. Perhaps it would help if we tried to write down some typing rules and either a reduction or CK semantics for this.

File: notes/fomega/evaluation/builtin-evaluation.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/fomega/evaluation/builtin-evaluation.md]: >  Indeed, it would also be nice if the reference implementations performed type substitutions rather than dropping them.

I shall do that in the typed CK machine, but do we want to perform type substitutions in the CEK machine as well?  Presumably we'd want to handle types analogously to variables, which I think would involve putting a type environment into the CEK monad and also into every CekValue.  That's entirely doable, but a bit of a pain.

[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/fomega/evaluation/builtin-evaluation.md]: > I shall do that in the typed CK machine, but do we want to perform type substitutions in the CEK machine as well? 

I think not now, but if we have the untyped machine as the "production" version, then I'd be quite happy to have the typed version take a performance hit and actually do the type substitutions as you say.

[the remark this answers is not in the document]
Comment (jmchapman) [reply notes/fomega/evaluation/builtin-evaluation.md]: I agree with both comments. Note that in the formalisation currently the CEK machine does ordinary type substitutions instead of delaying them as it does for term substitutions. So, I guess it's a bit of a hybrid between the CEK and CK machine. I should probably do it properly as Kenneth suggests.
